scalyr_agent/copying_manager/worker.py

```diff
@@ -1021,29 +1033,53 @@ def __scan_for_new_bytes(self, current_time=None):
         for processor in self.__log_processors:
             processor.scan_for_new_bytes(current_time)
 
-    def __write_checkpoint_state(
-        self, log_processors, base_file, current_time, full_checkpoint
-    ):
-        # type: (List[LogFileProcessor], six.text_type, float, bool) -> None
-        """Writes the current checkpoint state to disk.
+    def _remove_closed_log_processors_and_save_checkpoints(self):
+        new_log_processors = []
 
-        This must be done periodically to ensure that if the agent process stops and starts up again, we pick up
-        from where we left off copying each file.
+        for processor in self.__log_processors:
+            if processor.is_closed():
+                self._closed_log_processors.append(processor)
+            else:
+                new_log_processors.append(processor)
+
+        self.__log_processors = new_log_processors
+
+        checkpoints = self._get_log_processors_checkpoints(
+            self._closed_log_processors, time.time()
+        )
+        with self._closed_files_checkpoints_lock:
+            self._closed_files_checkpoints.update(checkpoints)
+
+    def get_closed_files_checkpoints(self):  # type: () -> Dict[six.text_type, Any]
```

Comment:
We should have a different name for this method since this mutates state.  Maybe `get_and_reset_closed_files_checkpoints`?  Or something like that.